File: iOS/DataPool/Sources/DataPool/Type/PoolName.swift
private enum PoolNameError: Error {
    case invalidLength
}

public struct PoolName {
    public let value: String
    
    public init?(_ value: String) {
        if !PoolName.isValid(value) {
            return nil
        }
        self.value = value
    }
    
    public static func isValid(_ value: String) -> Bool {
        do {
            try checkLength(value)
            return true
        } catch {
            return false
        }
    }
    
    private static func checkLength(_ value: String) throws {
        if !(!value.isEmpty && value.count <= 100) {
            throw PoolNameError.invalidLength
        }
    }
}

File: Android/data/pool/src/main/java/com/felipearpa/tyche/data/pool/type/PoolName.kt
package com.felipearpa.tyche.data.pool.type

@JvmInline
value class PoolName(val value: String) {
    init {
        check(value)
    }

    override fun toString(): String = value

    companion object {
        fun isValid(value: String): Boolean {
            return try {
                check(value = value)
                true
            } catch (ignored: IllegalArgumentException) {
                false
            }
        }

        private fun check(value: String) {
            checkLength(value)
        }

        private fun checkLength(value: String) {
            require(!(value.isNotEmpty() && value.length <= 100)) { "Length of pool name must be between 1 and 100" }
        }
    }
}
